src/interfaces/libpq/fe-connect.c

```diff
@@ -3523,6 +3529,10 @@ internal_cancel(SockAddr *raddr, int be_pid, int be_key,
 	 */
 #ifndef WIN32
 retry5:
+
+	if (bypass_conn_close_at_cancel)
+		goto cancel_errReturn;
```

Comment:
why not just ignore gpdb specific block?
```suggestion
		goto retry6;
```

Reply:
we can't go to `retry6`, as in this case we will hang on `recv`.

Reply:
has pure postgres same error?

Reply:
Libpq in the pure postgres behaves similarly (with the difference that it doesn't have `poll`, so it hangs on `recv`; and the logic of `internal_cancel` resides directly in `PQcancel`).
Of course, we can't have the same use case in pure postgres to check. The most similar and simple scenario I checked is:
1. Connect to running db via psql;
2. Send stop signal to backend and postmaster;
3. Launch a simple query (like `select from t;`) in the psql. The query will be waiting for the db to return results;
4. Send the `SIGINT` to the psql. The `SIGINT` should cancel current query, it has a `PQcancel` in the signal handler;
5. The query is not canceled. The stack is following, as we can see it hangs on `recv` till the CONT signal is sent to the backend:  

```
	__libc_recv() at recv.c:28 0x7f8d52727670	
	PQcancel() at 0x7f8d52a5bb0e	
	handle_sigint() at 0x5595860aaabe	
	<signal handler called>() at 0x7f8d52642520	
	__GI___poll() at poll.c:29 0x7f8d52718b97	
	PQsocketPoll() at 0x7f8d52a69a1a	
	pqSocketCheck() at 0x7f8d52a69d93	
	pqWaitTimed() at 0x7f8d52a69e64	
	PQgetResult() at 0x7f8d52a661a8	
	ExecQueryAndProcessResults() at 0x559586077817	
	SendQuery() at 0x5595860763ff	
	MainLoop() at 0x55958608b264	
	main() at 0x55958606cae3
```